experimental/Schemes/WeilDivisor.jl

```diff
@@ -0,0 +1,352 @@
+  R === coefficient_ring(E) || error("coefficient rings do not coincide")
+  dict = IdDict{IdealSheaf, elem_type(R)}()
+  for I in keys(coefficient_dict(D))
+    dict[I] = D[I]
+  end
+  for I in keys(coefficient_dict(E))
+    if haskey(dict, I)
+      c = D[I] + E[I]
+      if iszero(c) 
+        delete!(dict, I)
+      else 
+        dict[I] = c
+      end
+    else
+      dict[I] = E[I]
+    end
+  end
+  return WeilDivisor(X, R, dict, check=false)
+end
+
+function -(D::T) where {T<:WeilDivisor}
+  dict = IdDict{IdealSheaf, elem_type(coefficient_ring(D))}()
+  for I in keys(coefficient_dict(D))
+    dict[I] = -D[I]
+  end
+  return WeilDivisor(scheme(D), coefficient_ring(D), dict, check=false)
+end
+
+-(D::T, E::T) where {T<:WeilDivisor} = D + (-E)
+
+function *(a::RingElem, E::WeilDivisor)
+  dict = IdDict{IdealSheaf, elem_type(coefficient_ring(E))}()
+  for I in keys(coefficient_dict(E))
+    c = a*E[I]
+    if iszero(c)
+      delete!(dict, I)
+    else
+      dict[I] = c
+    end
+  end
+  return WeilDivisor(scheme(E), coefficient_ring(E), dict, check=false)
+end
+
+*(a::Int, E::WeilDivisor) = coefficient_ring(E)(a)*E
+*(a::Integer, E::WeilDivisor) = coefficient_ring(E)(a)*E
+
++(D::WeilDivisor, I::IdealSheaf) = D + WeilDivisor(I)
+
+function ==(D::WeilDivisor, E::WeilDivisor) 
+  keys(coefficient_dict(D)) == keys(coefficient_dict(E)) || return false
+  for I in keys(coefficient_dict(D))
+    if haskey(coefficient_dict(E), I)
+      D[I] == E[I] || return false
+    else
+      iszero(D[I]) || return false
+    end
+  end
+  for I in keys(coefficient_dict(E))
+    !(I in keys(coefficient_dict(D))) && !(iszero(E[I])) && return false
+  end
+  return true
+end
+
+#function intersection(D::T, E::T) where {T<:WeilDivisor}
+#  X = scheme(D)
+#  X == scheme(E) || error("divisors do not live on the same scheme")
+#  R = coefficient_ring(D)
+#  R == coefficient_ring(E) || error("divisors do not have the same coefficient ring")
+#  # prepare a copy of the divisors
+#  D_copy = WeilDivisor(X, R)
+#  E_copy = WeilDivisor(X, R)
+#  # check whether a common refinement of the covering is necessary
+#  CD = covering(D)
+#  CE = covering(E)
+#  if CD != CE
+#    CC, f, g = common_refinement(X, CD, CE)
+#    D_copy = pullback(f, D)
+#    E_copy = pullback(g, E)
+#  else
+#    D_copy = D
+#    E_copy = E
+#  end
+#  # TODO: Work out the intersection
+#end
+
+@Markdown.doc """
+    in_linear_system(f::VarietyFunctionFieldElem, D::WeilDivisor; check::Bool=true)
+
+Returns `true` if the rational function `f` is in the linear system ``|D|``
+and `false` otherwise.
+"""
+function in_linear_system(f::VarietyFunctionFieldElem, D::WeilDivisor; check::Bool=true)
+  X = scheme(D) 
+  X === variety(parent(f)) || error("schemes not compatible")
+  C = default_covering(X)
+  for I in components(D)
+    order_on_divisor(f, I, check=check) >= -D[I] || return false
+  end
+  for U in patches(C)
+    # we have to check that f[U] has no poles outside D[U]
```

Comment:
You need to find a single affine open where `f` is regular. 
Then the poles lie in the complement which is a proper closed subscheme `Z`. 
Since `X` is noetherian there are only finitely many divisors contained in `Z`.